src/coreComponents/mesh/generators/ParMETISInterface.cpp

```diff
@@ -0,0 +1,107 @@
+#include "common/GEOS_RAJA_Interface.hpp"
+
+#include <parmetis.h>
+
+#include <numeric>
+
+namespace geosx
+{
+namespace parmetis
+{
+
+static_assert( std::is_same< idx_t, int64_t >::value, "PetMETIS must be built with IDXTYPEWIDTH=64" );
+
+array1d< int64_t >
+partMeshKway( ArrayOfArraysView< int64_t const, int64_t > const & elemToNodes,
+              MPI_Comm comm,
+              int const minCommonNodes,
+              int const numRefinements )
+{
+  int const numProcs = MpiWrapper::commSize( comm );
+  idx_t const numElems = elemToNodes.size();
+
+  array1d< int64_t > part( numElems ); // all 0 by default
+  if( numProcs > 1 )
+  {
+    // Compute `elemdist` parameter (element range owned by each rank)
+    array1d< idx_t > elemCounts;
+    MpiWrapper::allGather( numElems, elemCounts, comm );
+    array1d< idx_t > elemDist( numProcs + 1 );
+    std::partial_sum( elemCounts.begin(), elemCounts.end(), elemDist.begin() + 1 );
+
+    // Compute tpwgts parameters (target partition weights)
+    array1d< real_t > tpwgts( numProcs );
+    tpwgts.setValues< serialPolicy >( static_cast< real_t >( 1.0 / numProcs ) );
+
+    // Set other ParMETIS parameters
+    idx_t wgtflag = 0;
+    idx_t numflag = 0;
+    idx_t ncon = 1;
+    idx_t ncommonnodes = minCommonNodes;
+    idx_t npart = numProcs;
+    idx_t options[4] = { 1, 0, 2022, PARMETIS_PSR_UNCOUPLED };
+    idx_t edgecut = 0;
+    real_t ubvec = 1.05;
+
+    auto const checkReturnValue = []( int const r, char const * func )
+    {
+      GEOSX_ERROR_IF_NE_MSG( r, METIS_OK, GEOSX_FMT( "ParMETIS call '{}' returned an error code: {}", func, r ) );
+    };
+
+    idx_t * xadj;
+    idx_t * adjncy;
+    {
+      int const res = ParMETIS_V3_Mesh2Dual( elemDist.data(),
+                                             const_cast< idx_t * >( elemToNodes.getOffsets() ),
+                                             const_cast< idx_t * >( elemToNodes.getValues() ),
+                                             &numflag, &ncommonnodes,
+                                             &xadj, &adjncy, &comm );
+      checkReturnValue( res, "Mesh2Dual" );
+    }
+
+    // Call ParMETIS
+    {
+      int const res = ParMETIS_V3_PartKway( elemDist.data(), xadj, adjncy,
+                                            nullptr, nullptr, &wgtflag,
+                                            &numflag, &ncon, &npart, tpwgts.data(),
+                                            &ubvec, options, &edgecut, part.data(), &comm );
+      checkReturnValue( res, "PartKway" );
+    }
+
+    for( int iter = 0; iter < numRefinements; ++iter )
+    {
+      int const res = ParMETIS_V3_RefineKway( elemDist.data(), xadj, adjncy,
+                                              nullptr, nullptr, &wgtflag,
+                                              &numflag, &ncon, &npart, tpwgts.data(),
+                                              &ubvec, options, &edgecut, part.data(), &comm );
+      checkReturnValue( res, "RefineKway" );
+    }
```

Comment:
I went back to the tests that we discussed (~6M cells, 64 ranks) and tested with:
- `partitionRefinement="1"`
- `partitionRefinement="2"`
- `partitionRefinement="3"`

I think that with  `partitionRefinement="1"` the quality is already significantly improved compared to the native VTK partitions. As `partitionRefinement` is increased, we see the improvement of the partitions, although the shape of the original VTK partitions is still visible for 3 (but it is not an issue at all). When the PR is merged, I will further test to see what happens if I choose 10, 20, etc

Reply:
Yes, I find that additional refinement iterations have very quickly diminishing returns. As soon as the imbalance tolerance limit is reached (hardcoded to 1.05 currently), ParMETIS does not seem eager to do too many changes. I do think that the partitions it produces look reasonable in most cases.